src/modules/ims_registrar_pcscf/service_routes.c

```diff
@@ -117,43 +121,79 @@ int checkcontact(struct sip_msg *_m, pcontact_t *c)
 {
 	int security_server_port = -1;
 	str received_host = {0, 0};
+	unsigned short received_port = 0;
+	char received_proto = 0;
 	char srcip[50];
 
+	if(trust_bottom_via) {
+		struct via_body *vb = cscf_get_last_via(_m);
+		if(vb == 0) {
+			LM_ERR("No via header in request\n");
+			return -1;
+		}
+		if(vb->received != NULL && vb->received->value.len > 0) {
+			received_host = vb->received->value;
+		} else {
+			received_host = vb->host;
+		}
+		if(vb->rport != NULL && vb->rport->value.len > 0) {
+			str2ushort(&vb->rport->value, &received_port);
+		} else {
+			received_port = vb->port;
+		}
+		// this probably doesn't matter, since IMS UEs register IPs for both
+		received_proto = vb->proto;
+	} else {
+		received_host.len = ip_addr2sbuf(&_m->rcv.src_ip, srcip, sizeof(srcip));
+		received_host.s = srcip;
+		received_port = _m->rcv.src_port;
+		received_proto = _m->rcv.proto;
+	}
+
 	LM_DBG("Port %d (search %d), Proto %d (search %d), reg_state %s (search "
 		   "%s)\n",
-			c->received_port, _m->rcv.src_port, c->received_proto,
-			_m->rcv.proto, reg_state_to_string(c->reg_state),
+			c->received_port, received_port, c->received_proto, received_proto,
+			reg_state_to_string(c->reg_state),
 			reg_state_to_string(PCONTACT_REGISTERED));
 
-	if(c->security) {
-		switch(c->security->type) {
-			case SECURITY_IPSEC:
-				security_server_port = c->security->data.ipsec->port_uc;
-				break;
-			case SECURITY_TLS:
-			case SECURITY_NONE:
-				break;
+
+	if(ipsec_pcscf.ipsec_on_expire == NULL) {
+		LM_DBG("ims_ipsec_pcscf module not loaded - skipping port-uc checks\n");
+	} else if(!ignore_contact_rxport_check) {
+		if(c->security) {
+			switch(c->security->type) {
+				case SECURITY_IPSEC:
+					security_server_port = c->security->data.ipsec->port_uc;
+					break;
+				case SECURITY_TLS:
+				case SECURITY_NONE:
+					break;
+			}
+		} else if(c->security_temp) {
+			switch(c->security_temp->type) {
+				case SECURITY_IPSEC:
+					security_server_port =
+							c->security_temp->data.ipsec->port_uc;
+					break;
+				case SECURITY_TLS:
+				case SECURITY_NONE:
+					break;
+			}
 		}
-	} else if(c->security_temp) {
-		switch(c->security_temp->type) {
-			case SECURITY_IPSEC:
-				security_server_port = c->security_temp->data.ipsec->port_uc;
-				break;
-			case SECURITY_TLS:
-			case SECURITY_NONE:
-				break;
+
+		if(c->received_port != received_port
+				&& security_server_port != received_port) {
+			LM_DBG("check contact failed - port-uc %d is neither contact "
+				   "received_port %d, nor message received port %d\n",
+					security_server_port, c->received_port, _m->rcv.src_port);
+			return 1;
 		}
 	}
 
 	if((ignore_reg_state || (c->reg_state == PCONTACT_REGISTERED))
-			&& (ignore_contact_rxport_check
-					|| (c->received_port == _m->rcv.src_port)
-					|| (security_server_port == _m->rcv.src_port))
-			&& (ignore_contact_rxport_check
-					|| (c->received_proto == _m->rcv.proto))) {
```

Comment:
For context - the `ignore_contact_rxproto_check` parameter fixes this logical condition group. I don't know if this was on purpose, or a copy-paste mistake (checking proto for port flag).